### PR TITLE
HPCC-26700 Report worker pod skew

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: common/workunit/workunit.cpp [hunk 1]
```diff
@@ -14467,4 +14467,55 @@ std::pair<std::string, unsigned> getExternalService(const char *serviceName)
     return servicePair;
 }
 
+std::vector<std::vector<std::string>> getPodNodes(const char *selector)
+{
+    VStringBuffer getWorkerNodes("kubectl get pods --selector=job-name=%s \"--output=jsonpath={range .items[*]}{.metadata.name},{.spec.nodeName}{'\\n'}{end}\"", selector);
+    StringBuffer result;
+    runKubectlCommand("get-workder-nodes", getWorkerNodes, nullptr, &result);
```
Comment (ghalliday) [inline common/workunit/workunit.cpp:14474]: get-worker-nodes

File: common/workunit/workunit.cpp [hunk 2]
```diff
@@ -14467,4 +14467,55 @@ std::pair<std::string, unsigned> getExternalService(const char *serviceName)
     return servicePair;
 }
 
+std::vector<std::vector<std::string>> getPodNodes(const char *selector)
```
Comment (ghalliday) [inline common/workunit/workunit.cpp:14470]: std:vector<std::pair<std:string, std:string>> seems more natural.  And using a typedef for the element.
Slightly concerned about the efficiency or reallocation (and using pair is possibly worse))

Comment (ghalliday) [reply common/workunit/workunit.cpp:14470]: Could do with a comment describing what the function returns.

File: thorlcr/master/thgraphmanager.cpp [hunk 3]
```diff
@@ -1001,12 +1002,52 @@ bool CJobManager::executeGraph(IConstWorkUnit &workunit, const char *graphName,
...
         {
+#ifdef _CONTAINERIZED
+            // collate pod distribution
+            VStringBuffer selector("thorworker-%s-%s", wuid.get(), graphName);
+            std::vector<std::vector<std::string>> pods = getPodNodes(selector.toLowerCase());
+            std::unordered_map<std::string, unsigned> podPerNodeCounts;
+            for (const auto &podNode: pods)
+            {
+                const std::string &node = podNode[1]; // pod is 1st item, node is 2nd
+                podPerNodeCounts[node]++; // NB: if doesn't exist is created with default value of 0 1st
```
Comment (ghalliday) [inline thorlcr/master/thgraphmanager.cpp:1013]: is getPodNodes likely to be called again?  If not it would be much more efficient for this code to be calculated directly instead of creating the intermediate vector

File: thorlcr/master/thgraphmanager.cpp [hunk 4]
```diff
@@ -1001,12 +1002,52 @@ bool CJobManager::executeGraph(IConstWorkUnit &workunit, const char *graphName,
...
+            double sumSquares=0;
+            for (auto &node: podPerNodeCounts)
+            {
+                podStats.setStatistic(StNumPods, node.second, nodes++);
+                sum += node.second;
+                sumSquares += node.second*node.second;
+            }
+            double mean = sum / nodes;
+            double variance = (sumSquares - sum * mean) / nodes;
+            double stdDev = sqrt(variance);
```
Comment (ghalliday) [inline thorlcr/master/thgraphmanager.cpp:1028]: sum and sumSquares are already calculated in podStats.  
Cleaner to add a CRuntimeSummaryStatisticCollection::queryStdDev(kind) function:
{ derived[queryMapping().getIndex(kind)]->queryStdDev() }

File: thorlcr/master/thgraphmanager.cpp [hunk 5]
```diff
@@ -1001,12 +1002,52 @@ bool CJobManager::executeGraph(IConstWorkUnit &workunit, const char *graphName,
...
+
+#ifdef _CONTAINERIZED
+            // issue warning and publish pod distribution stats, if any stddev
+            if (stdDev)
+            {
+                Owned<IStatisticGatherer> collector = createGlobalStatisticGatherer(wu);
+                StatsScopeId graphScopeId(graphName);
+                collector->beginScope(graphScopeId);
+                podStats.recordStatistics(*collector, false);
+                Owned<IException> e = makeStringExceptionV(-1, "Degraded performance. Worker pods are unevenly distributed over nodes. StdDev=%.2f. See workunit stats. for more info.", stdDev);
```
Comment (ghalliday) [inline thorlcr/master/thgraphmanager.cpp:1047]: This doesn't tell you which slave pods were on the nodes that had fewer or more pods.  Is that going to be something you wish you knew?

**Review by ghalliday — COMMENTED** [review]
A few questions.

**Review by jakesmith — COMMENTED** [review]
@ghalliday - please see changes.

**Review by ghalliday — COMMENTED** [review]
@jakesmith

**Review by jakesmith — COMMENTED** [review]
@ghalliday - please see new commit changes.

File: thorlcr/master/thgraphmanager.cpp [hunk 6]
```diff
@@ -1039,10 +1048,17 @@ bool CJobManager::executeGraph(IConstWorkUnit &workunit, const char *graphName,
             if (stdDev)
             {
                 Owned<IStatisticGatherer> collector = createGlobalStatisticGatherer(wu);
-                StatsScopeId graphScopeId(graphName);
+                StatsScopeId wfidScopeId(SSTworkflow, wfid);
+                StatsScopeId graphScopeId;
```
Comment (ghalliday) [inline thorlcr/master/thgraphmanager.cpp:1052]: trivial: Why the change from 
`StatsScopeId graphScopeId(graphName)`

**Review by ghalliday — APPROVED** [review]
@jakesmith I think it is ok, you may want to address my last two comments.  Squash regardless.

File: thorlcr/master/thgraphmanager.cpp [hunk 7]
```diff
@@ -1001,12 +1002,50 @@ bool CJobManager::executeGraph(IConstWorkUnit &workunit, const char *graphName,
...
+#ifdef _CONTAINERIZED
+            // issue warning and publish pod distribution stats, if any stddev
+            if (stdDev)
+            {
+                Owned<IStatisticGatherer> collector = createGlobalStatisticGatherer(wu);
+                StatsScopeId graphScopeId(graphName);
+                collector->beginScope(graphScopeId);
+                podStats.recordStatistics(*collector, false);
+                Owned<IException> e = makeStringExceptionV(-1, "Degraded performance. Worker pods are unevenly distributed over nodes. StdDev=%.2f. min node(%s) has %" I64F "u pods, max node(%s) has %" I64F "u pods", stdDev, nodeNames[minNode].c_str(), min, nodeNames[maxNode].c_str(), max);
+                reportExceptionToWorkunit(*wu, e);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply thorlcr/master/thgraphmanager.cpp:1062]: hm yes, IWUException::setScope. That's not being used anywhere by Thor at the moment.
It doesn't look like many other places use/set it either.
All activity/graphs exception (which already add text via a common method), could probably call setScope on reported exceptions.
I've opened a separate JIRA to revisit the issue of using IWUException::setScope : https://track.hpccsystems.com/browse/HPCC-26768

File: thorlcr/master/thgraphmanager.cpp (hunk 6 above)
Comment (jakesmith) [reply thorlcr/master/thgraphmanager.cpp]: not sure, reverted.

**Review by jakesmith — COMMENTED** [review]
@ghalliday - I think this is ready to merge now.

File: common/workunit/workunit.hpp [hunk 8]
```diff
@@ -1768,6 +1768,8 @@ extern WORKUNIT_API void runK8sJob(const char *componentName, const char *wuid,
 
 // return the k8s external host and port for serviceName
 extern WORKUNIT_API std::pair<std::string, unsigned> getExternalService(const char *serviceName);
+
 #endif
+extern WORKUNIT_API std::vector<std::vector<std::string>> getPodNodes(const char *selector);
```
Comment (ghalliday) [inline common/workunit/workunit.hpp:1773]: trivial: Should be inside the #endif above.

**Review by ghalliday — APPROVED** [review]

File: common/workunit/workunit.hpp (hunk 8 above)
Comment (jakesmith) [reply common/workunit/workunit.hpp:1773]: fixed + squashed

**Review by ghalliday — APPROVED** [review]
